sockeye/inference.py

```diff
@@ -862,17 +912,22 @@ def _get_inference_input(self,
         for j, trans_input in enumerate(trans_inputs):
             num_tokens = len(trans_input)  # includes eos
             max_output_lengths.append(self._get_max_output_length(num_tokens))
-            source[j, :num_tokens, 0] = tokens2ids(trans_input.tokens, self.source_vocabs[0])
-
+            num_tokens = num_tokens + trans_input.num_source_prefix_tokens()
+            source[j, :num_tokens, 0] = tokens2ids(trans_input.get_source_prefix_tokens() \
+                + trans_input.tokens, self.source_vocabs[0])
             factors = trans_input.factors if trans_input.factors is not None else []
             num_factors = 1 + len(factors)
             if num_factors != self.num_source_factors:
                 logger.warning("Input %d factors, but model(s) expect %d", num_factors,
                                self.num_source_factors)
-            for i, factor in enumerate(factors[:self.num_source_factors - 1], start=1):
-                # fill in as many factors as there are tokens
-
-                source[j, :num_tokens, i] = tokens2ids(factor, self.source_vocabs[i])[:num_tokens]
+            if not trans_input.source_prefix_factors: # no source prefix during inference
+                for i, factor in enumerate(factors[:self.num_source_factors - 1], start=1):
+                    # fill in as many factors as there are tokens
+                    source[j, :num_tokens, i] = tokens2ids(factor, self.source_vocabs[i])[:num_tokens]
+            else:
+                for i, zip_of_factor_and_prefix_factor in enumerate(zip(factors[:self.num_source_factors - 1], trans_input.source_prefix_factors[:self.num_source_factors - 1]), start=1):
+                    factor, source_prefix_factor = zip_of_factor_and_prefix_factor
+                    source[j, :num_tokens, i] = tokens2ids(source_prefix_factor + factor, self.source_vocabs[i])[:num_tokens]
```

Comment:
we could avoid the explicit construction of the concatenated list with `chain(source_prefix_factor, factor)`

Reply:
done!